### PR TITLE
feat(k8s-manager): add horizontal pod autoscaler when multiple replicas

### PR DESCRIPTION
**Pull request by JYisus**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other changes (ci configuration, documentation or any other kind of changes)

## Checklist:

- [x] I have created tests for my code changes, and the tests are passing.
- [x] I have executed the pre-commit hooks locally.
- [ ] My change requires a change to the documentation (create a new issue if the documentation has not been updated).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: engine/k8s-manager/internal/infrastructure/kube/process/create.go [hunk 1]
```diff
@@ -32,29 +31,44 @@ func (kp *KubeProcess) Create(
...
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      processIdentifier,
 			Namespace: kp.namespace,
 			Labels:    labels,
 		},
 		Spec: appsv1.DeploymentSpec{
-			Replicas: pointer.Int32(spec.Process.Replicas),
```
Comment (alexrodfe) [inline engine/k8s-manager/internal/infrastructure/kube/process/create.go:57]: watchout dead code

File: engine/k8s-manager/internal/infrastructure/kube/process/create.go [hunk 2]
```diff
@@ -32,29 +31,44 @@ func (kp *KubeProcess) Create(
...
 }
 
 func (kp *KubeProcess) getDeploymentSpec(configMapName string, spec *processSpec) *appsv1.Deployment {
 	labels := kp.getProcessLabels(spec)
 
-	processIdentifier := getFullProcessIdentifier(spec.Product, spec.Version, spec.Workflow, spec.Process.Name)
+	processIdentifier := getDeploymentName(spec.Product, spec.Version, spec.Workflow, spec.Process.Name)
 
 	return &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
```
Comment (alexrodfe) [inline engine/k8s-manager/internal/infrastructure/kube/process/create.go:61]: out of curiosity, why is this important to add?

File: engine/k8s-manager/internal/infrastructure/kube/process/create.go [hunk 3]
```diff
@@ -32,29 +31,44 @@ func (kp *KubeProcess) Create(
...
+			APIVersion: "apps/v1",
+		},
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      processIdentifier,
 			Namespace: kp.namespace,
 			Labels:    labels,
 		},
 		Spec: appsv1.DeploymentSpec{
-			Replicas: pointer.Int32(spec.Process.Replicas),
+			//Replicas: pointer.Int32(spec.Process.Replicas),
```
Comment (alexrodfe) [inline engine/k8s-manager/internal/infrastructure/kube/process/create.go:71]: watchout dead code

File: engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden [hunk 4]
```diff
@@ -33,7 +33,7 @@ items:
         zzz_deprecatedclustername: ""
         managedfields: []
       spec:
-        replicas: 1
+        replicas: null
```
Comment (alexrodfe) [inline engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden:36]: is desirable to have null replicas as an example?

**Review by alexrodfe — CHANGES_REQUESTED** [review]
Great!
I have a couple of questions before proceeding

File: engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden (hunk 4 above)
Comment (alexrodfe) [reply engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden:36]: note: all other changed goldens have changed their replicas to null

File: engine/k8s-manager/internal/infrastructure/kube/process/autoscaler.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+			MaxReplicas: process.Replicas,
+			Metrics: []autoscalilngv2.MetricSpec{
+				{
+					Type: autoscalilngv2.ContainerResourceMetricSourceType,
+					ContainerResource: &autoscalilngv2.ContainerResourceMetricSource{
+						Name:      corev1.ResourceCPU,
+						Container: process.Name,
+						Target: autoscalilngv2.MetricTarget{
+							Type:               autoscalilngv2.UtilizationMetricType,
+							AverageUtilization: pointer.Int32(80),
```
Comment (dfinteligenz) [inline engine/k8s-manager/internal/infrastructure/kube/process/autoscaler.go:49]: I'd retrieve the average utilization from viper, and set a default value for the property of 80.
This way we can override it with external configuration (env variable or configuration file) if needed.

File: engine/k8s-manager/internal/infrastructure/kube/process/create.go (hunk 2 above)
Comment (dfinteligenz) [reply engine/k8s-manager/internal/infrastructure/kube/process/create.go:61]: We thought the Kubernetes API was setting the values automatically, but it wasn't, and when implementing the HPA it was problematic.

File: engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden (hunk 4 above)
Comment (dfinteligenz) [reply engine/k8s-manager/internal/infrastructure/kube/process/testdata/StartProcess.golden:36]: The replicas field on the deployment is not needed anymore.

For upscaling to N replicas, the replica number is defined in the HPA and the HPA is the responsible of upscaling when needed based on the specified metrics.

**Review by dfinteligenz — CHANGES_REQUESTED** [review]

**Review by dfinteligenz — APPROVED** [review]
Great job!

**Pull request merged**